### PR TITLE
Update prometheus-cpp and Metric API unit tests

### PR DESCRIPTION
**Pull request by rmccorm4**
- [x] Update Prometheus API usage for new prometheus-cpp version for metter support in managing duplicate metrics/metric families. Before this bump:
    - Duplicately named metric families of different types (counter, gauge, etc.) could be added to the registry which is against the prometheus spec
    - Duplicately named metric families (even of same type) would keep being added as separate objects to registry. Now duplicates will return a reference to the same family.
- [x] Update unit tests to more thoroughly test duplicates and edge cases
- [x] Think about out-of-order deletion: i.e. calling `MetricFamilyDelete()` before `MetricDelete()` where `Metric->family_->Remove()` reference is called in destructor even though `family_` should be invalid memory now.
- [x] Check if existing metric families from core (nv_request..., nv_gpu_util..., etc.) can be inadvertently removed by creating a duplicate MF with Metrics API and then calling registry->Remove() on it in destructor
    - it can, delaying registry removal until explicitly asked for, see DLIS-4072

### CODE REVIEW

**Comment by rmccorm4** [comment]
New unit tests:
```
[==========] Running 9 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 9 tests from MetricsApiTest
[ RUN      ] MetricsApiTest.TestCounterEndToEnd
[       OK ] MetricsApiTest.TestCounterEndToEnd (241 ms)
[ RUN      ] MetricsApiTest.TestGaugeEndToEnd
[       OK ] MetricsApiTest.TestGaugeEndToEnd (1 ms)
[ RUN      ] MetricsApiTest.TestDupeMetricFamilyDiffKind
[       OK ] MetricsApiTest.TestDupeMetricFamilyDiffKind (0 ms)
[ RUN      ] MetricsApiTest.TestDupeMetricFamilyDiffDescription
[       OK ] MetricsApiTest.TestDupeMetricFamilyDiffDescription (1 ms)
[ RUN      ] MetricsApiTest.TestDupeMetricFamily
[       OK ] MetricsApiTest.TestDupeMetricFamily (1 ms)
[ RUN      ] MetricsApiTest.TestDupeMetricLabels
metric1 value: 7.5 == metric2 value: 7.5
[       OK ] MetricsApiTest.TestDupeMetricLabels (0 ms)
[ RUN      ] MetricsApiTest.TestDupeMetricEmptyLabels
metric1 value: 7.5 == metric2 value: 7.5
[       OK ] MetricsApiTest.TestDupeMetricEmptyLabels (1 ms)
[ RUN      ] MetricsApiTest.TestOutOfOrderDelete
[       OK ] MetricsApiTest.TestOutOfOrderDelete (0 ms)
[ RUN      ] MetricsApiTest.TestCoreMetricAccess
[       OK ] MetricsApiTest.TestCoreMetricAccess (0 ms)
[----------] 9 tests from MetricsApiTest (250 ms total)

[----------] Global test environment tear-down
[==========] 9 tests from 1 test suite ran. (250 ms total)
[  PASSED  ] 9 tests.
```

File: src/test/metrics_api_test.cc [hunk 1]
```diff
@@ -236,30 +344,239 @@ TEST_F(MetricsApiTest, TestGaugeEndToEnd)
...
+  const char* description = "diff kind description";
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_MetricFamilyNew(&family1, kind1, name, description),
+      "Creating new metric family1");
+
+  // Create duplicate metric family with different kind
+  TRITONSERVER_MetricFamily* family2 = nullptr;
+  TRITONSERVER_MetricKind kind2 = TRITONSERVER_METRIC_KIND_GAUGE;
+  // Expect this to fail, can't have duplicate name of different kind
+  auto err = TRITONSERVER_MetricFamilyNew(&family2, kind2, name, description);
```
Comment (Tabrizian) [inline src/test/metrics_api_test.cc:376]: If the user's Python implementation have two instances of the model and they are calling `TRITONSERVER_MetricFamilyNew` in their `initialize` function, the initialize call for the second model instance will fail since the metric family already exists. Do you think it makes sense to return the existing metric family instead of returning an error?

File: src/test/metrics_api_test.cc [hunk 2]
```diff
@@ -236,30 +344,239 @@ TEST_F(MetricsApiTest, TestGaugeEndToEnd)
...
+
+  std::vector<const TRITONSERVER_Parameter*> labels2;
+  labels2.emplace_back(TRITONSERVER_ParameterNew(
+      metric_key, TRITONSERVER_PARAMETER_STRING, "label2"));
+  TRITONSERVER_Metric* metric2 = nullptr;
+  FAIL_TEST_IF_ERR(
+      TRITONSERVER_MetricNew(&metric2, family2, labels2.data(), labels2.size()),
+      "Creating new metric2");
+  for (const auto label : labels2) {
+    TRITONSERVER_ParameterDelete(const_cast<TRITONSERVER_Parameter*>(label));
```
Comment (Tabrizian) [inline src/test/metrics_api_test.cc:480]: I think it would be good to update the backend API documentation to mention that `TRITONSERVER_MetricNew` does not take the ownership of labels and the caller must delete them.

File: src/tritonserver.cc [hunk 3]
```diff
@@ -2962,8 +2960,16 @@ TRITONSERVER_Error*
 TRITONSERVER_MetricDelete(TRITONSERVER_Metric* metric)
 {
 #ifdef TRITON_ENABLE_METRICS
-  delete reinterpret_cast<tc::Metric*>(metric);
-  return nullptr;  // Success
+  auto lmetric = reinterpret_cast<tc::Metric*>(metric);
+  if (lmetric->Family() == nullptr) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INTERNAL,
+        "MetricFamily reference was invalidated before Metric was deleted. "
```
Comment (Tabrizian) [inline src/tritonserver.cc:2967]: I think we should update the documentation for the metrics API to mention that it is best to delete all the metrics corresponding to a metric family before deleting the metric family. Otherwise, an error object will be received.

File: src/test/metrics_api_test.cc (hunk 1 above)
Comment (rmccorm4) [reply src/test/metrics_api_test.cc:376]: If two pb instances call `MetricFamilyNew(same_name, same_kind, ...)`, it will succeed and they will get the exiting metric family back.

This test is checking if they try to do `family1 = MetricFamilyNew(same_name, kind=COUNTER)` and `family2 = MetricFamilyName(same_name, kind=GAUGE)`

It is not allowed to have duplicate names of different `kind`

File: src/tritonserver.cc (hunk 3 above)
Comment (rmccorm4) [reply src/tritonserver.cc:2967]: Great suggestion, done

File: include/triton/core/tritonserver.h [hunk 4]
```diff
@@ -2267,6 +2267,10 @@ TRITONSERVER_DECLSPEC TRITONSERVER_Error* TRITONSERVER_MetricFamilyNew(
     const char* name, const char* description);
 
 /// Delete a metric family object.
+/// A TRITONSERVER_MetricFamily* object should be deleted AFTER its
+/// corresponding TRITONSERVER_Metric* objects have been deleted.
+/// If a family is deleted before its metrics, an error will be
```
Comment (GuanLuo) [inline include/triton/core/tritonserver.h:2272]: If the family has been deleted, what happened if user calls other `TRITONSERVER_MetricXxx` functions? Should also return error as the metric will no longer be valid?

File: src/metric_family.cc [hunk 5]
```diff
@@ -62,62 +62,71 @@ MetricFamily::MetricFamily(
...
     }
     default:
       throw std::invalid_argument(
           "Unsupported family kind passed to Metric constructor.");
   }
 
-  std::lock_guard<std::mutex> lk(mtx_);
-  ++metric_ref_cnt_[metric];
-  return metric;
+  std::lock_guard<std::mutex> lk(metric_mtx_);
```
[the remark this answers is not in the document]
Comment (rmccorm4) [reply src/metric_family.cc:88]: Just to check if I understand roughly what you're going for, you're suggesting something along these lines?
```cpp
MetricFamily::MetricFamily() {
  ...
  std::lock_guard<std::mutex> lk(metric_mtx);
  auto prom_metric_sp = std::shared_ptr<void>(prom_metric, MyCustomDeleter);
  return prom_metric_sp;
}
```

```cpp
MetricFamily::Remove(std::shared_ptr<void> prom_metric_sp) {
  ...
  delete prom_metric_sp;
}
```

```cpp
void MetricFamily::MyCustomDeleter() {
  switch (kind_) {
    case TRITONSERVER_METRIC_KIND_COUNTER: {
      auto counter_family_ptr =
          reinterpret_cast<prometheus::Family<prometheus::Counter>*>(family_);
      auto counter_ptr = reinterpret_cast<prometheus::Counter*>(prom_metric);
      counter_family_ptr->Remove(counter_ptr);
      break;
    }
    case TRITONSERVER_METRIC_KIND_GAUGE: {
      auto gauge_family_ptr =
          reinterpret_cast<prometheus::Family<prometheus::Gauge>*>(family_);
      auto gauge_ptr = reinterpret_cast<prometheus::Gauge*>(prom_metric);
      gauge_family_ptr->Remove(gauge_ptr);
      break;
    }
    default:
      // Invalid kind should be caught in constructor
      LOG_ERROR << "Unsupported kind in Metric destructor.";
      break;
  }
}
```

CC @GuanLuo who added the ref counting

File: src/test/metrics_api_test.cc (hunk 1 above)
Comment (Tabrizian) [reply src/test/metrics_api_test.cc:376]: Got it. Thanks for the clarification. Resolving.

File: src/metric_family.cc [hunk 6]
```diff
@@ -62,62 +62,71 @@ MetricFamily::MetricFamily(
...
+  child_metrics_.insert(metric);
+  return prom_metric;
 }
 
 void
-MetricFamily::Remove(void* metric)
+MetricFamily::Remove(void* prom_metric, Metric* metric)
 {
+  // Remove reference to dependent Metric object
+  child_metrics_.erase(metric);
```
Comment (GuanLuo) [inline src/metric_family.cc:98]: lock?

File: include/triton/core/tritonserver.h (hunk 4 above)
Comment (rmccorm4) [reply include/triton/core/tritonserver.h:2272]: Adding more unit tests around this

Comment (rmccorm4) [reply include/triton/core/tritonserver.h:2272]: Technically MetricSet/Increment/Value etc. will work as they operate directly on the metric and don't need to communicate with tc::MetricFamily object. 

Since we don't remove a MetricFamily's prometheus::Family from the prometheus::Registry on ~MetricFamily destruction, the tc::Metric will have a valid reference to the prometheus::Metric still in the prometheus::registry and can update it accordingly.

If we want to disallow this, we can also invalidate the reference by setting `tc::Metric::metric_ = nullptr`? Let me know what you think.

Comment (rmccorm4) [reply include/triton/core/tritonserver.h:2272]: I also added a similar check to MetricFamilyDelete to instead enforce that FamilyDelete is called only once its dependent metrics are deleted.

File: src/metric_family.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (Tabrizian) [reply src/metric_family.cc:88]: It looks like it might not be easy to implement in this scenario. For me the original `ResponseFactory` was a shared pointer.

**Review by Tabrizian — APPROVED** [review]

**Review by GuanLuo — APPROVED** [review]
